File: frontend-web/src/services/contentService.ts
import apiService from './api';

interface Content {
  id: string;
  title: string;
  content_type: string;
  content_data: string;
  course_id: string;
  chapter_number: number;
  section_number: number;
  is_available: boolean;
  created_at: string;
  updated_at: string;
}

class ContentService {
  /**
   * Get all content for a specific course
   */
  public async getContentByCourse(courseId: string): Promise<Content[]> {
    try {
      const response = await apiService.get<Content[]>(`/content/by-course/${courseId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching content by course:', error);
      throw error;
    }
  }

  /**
   * Get specific content by ID
   */
  public async getContentById(contentId: string): Promise<Content> {
    try {
      const response = await apiService.get<Content>(`/content/${contentId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching content by ID:', error);
      throw error;
    }
  }

  /**
   * Search content by keywords
   */
  public async searchContent(query: string, courseId?: string): Promise<Content[]> {
    try {
      const params = new URLSearchParams({ q: query });
      if (courseId) params.append('course_id', courseId);
      
      const response = await apiService.get<Content[]>(`/content/search?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Error searching content:', error);
      throw error;
    }
  }

  /**
   * Get next content in sequence
   */
  public async getNextContent(contentId: string): Promise<Content> {
    try {
      const response = await apiService.get<Content>(`/content/${contentId}/next`);
      return response.data;
    } catch (error) {
      console.error('Error fetching next content:', error);
      throw error;
    }
  }

  /**
   * Get previous content in sequence
   */
  public async getPrevContent(contentId: string): Promise<Content> {
    try {
      const response = await apiService.get<Content>(`/content/${contentId}/prev`);
      return response.data;
    } catch (error) {
      console.error('Error fetching previous content:', error);
      throw error;
    }
  }

  /**
   * Get content for grounded Q&A
   */
  public async getGroundedQAContent(query: string, courseId?: string): Promise<Content[]> {
    try {
      const params = new URLSearchParams({ query });
      if (courseId) params.append('course_id', courseId);
      
      const response = await apiService.post<Content[]>('/content/grounded-qna', params);
      return response.data;
    } catch (error) {
      console.error('Error fetching grounded QA content:', error);
      throw error;
    }
  }
}

export default new ContentService();

File: backend-phase1-zero-backend-llm/src/services/content_service.py
from typing import List, Optional
from sqlalchemy.orm import Session
from sqlalchemy import desc
from ..models.course_content import CourseContent
from ..core.cloudflare_r2 import r2_client


class ContentService:
    """
    Content service that provides deterministic content delivery without any processing.
    All content is served verbatim from the database as stored, with no transformations
    or intelligent processing applied.
    """
    def __init__(self, db: Session):
        self.db = db

    def get_content_by_id(self, content_id: int) -> Optional[CourseContent]:
        """Retrieve course content by its ID"""
        return self.db.query(CourseContent).filter(CourseContent.id == content_id).first()

    def get_content_by_course(self, course_id: str, skip: int = 0, limit: int = 100) -> List[CourseContent]:
        """Retrieve all content for a specific course, ordered by chapter and section"""
        return (
            self.db.query(CourseContent)
            .filter(CourseContent.course_id == course_id)
            .filter(CourseContent.is_available == True)
            .order_by(CourseContent.chapter_number, CourseContent.section_number)
            .offset(skip)
            .limit(limit)
            .all()
        )

    def search_content(self, query: str, course_id: Optional[str] = None, skip: int = 0, limit: int = 10) -> List[CourseContent]:
        """Search course content by keywords"""
        search_query = self.db.query(CourseContent).filter(
            CourseContent.content_data.contains(query) | CourseContent.title.contains(query)
        ).filter(CourseContent.is_available == True)

        if course_id:
            search_query = search_query.filter(CourseContent.course_id == course_id)

        return search_query.offset(skip).limit(limit).all()

    def get_next_content(self, current_content: CourseContent) -> Optional[CourseContent]:
        """Get the next content in the course sequence"""
        next_content = (
            self.db.query(CourseContent)
            .filter(CourseContent.course_id == current_content.course_id)
            .filter(CourseContent.chapter_number == current_content.chapter_number)
            .filter(CourseContent.section_number > current_content.section_number)
            .filter(CourseContent.is_available == True)
            .order_by(CourseContent.section_number)
            .first()
        )

        if not next_content:
            # If no next section in current chapter, get first section of next chapter
            next_content = (
                self.db.query(CourseContent)
                .filter(CourseContent.course_id == current_content.course_id)
                .filter(CourseContent.chapter_number > current_content.chapter_number)
                .filter(CourseContent.is_available == True)
                .order_by(CourseContent.chapter_number, CourseContent.section_number)
                .first()
            )

        return next_content

    def get_prev_content(self, current_content: CourseContent) -> Optional[CourseContent]:
        """Get the previous content in the course sequence"""
        prev_content = (
            self.db.query(CourseContent)
            .filter(CourseContent.course_id == current_content.course_id)
            .filter(CourseContent.chapter_number == current_content.chapter_number)
            .filter(CourseContent.section_number < current_content.section_number)
            .filter(CourseContent.is_available == True)
            .order_by(desc(CourseContent.section_number))
            .first()
        )

        if not prev_content:
            # If no previous section in current chapter, get last section of previous chapter
            prev_content = (
                self.db.query(CourseContent)
                .filter(CourseContent.course_id == current_content.course_id)
                .filter(CourseContent.chapter_number < current_content.chapter_number)
                .filter(CourseContent.is_available == True)
                .order_by(desc(CourseContent.chapter_number), desc(CourseContent.section_number))
                .first()
            )

        return prev_content